packages/legend-query-builder/src/stores/milestoning/QueryBuilderMilestoningHelper.ts

```diff
@@ -0,0 +1,377 @@
+    }
+  }
+  return false;
+};
+
+/**
+ * Gets the value of ValueSpecification given INTERNAL__PropagatedValue is pointing to.
+ */
+export const getValueOfInternalPropagatedValue = (
+  valueSpec: INTERNAL__PropagatedValue,
+): ValueSpecification => {
+  if (valueSpec.getValue() instanceof INTERNAL__PropagatedValue) {
+    return getValueOfInternalPropagatedValue(
+      guaranteeType(valueSpec.getValue(), INTERNAL__PropagatedValue),
+    );
+  }
+  return valueSpec.getValue();
+};
+
+/**
+ * Check if the parameter value of the milestoned property is
+ * the same as those specified in global scope, so that we can
+ * potentially replace them with propgated value.
+ */
+export const matchMilestoningParameterValue = (
+  stereotype: MILESTONING_STEREOTYPE,
+  idx: number,
+  parameterValue: ValueSpecification,
+  milestoningDate: QueryBuilderMilestoningState,
+): boolean => {
+  const checkIfEquivalent = (
+    param1: ValueSpecification | undefined,
+    param2: ValueSpecification | undefined,
+  ): boolean => {
+    if (
+      param1 instanceof VariableExpression &&
+      param2 instanceof VariableExpression
+    ) {
+      return param1.name === param2.name;
+    } else if (
+      param1 instanceof PrimitiveInstanceValue &&
+      param2 instanceof PrimitiveInstanceValue
+    ) {
+      if (
+        param1.genericType.value.rawType.name === PRIMITIVE_TYPE.LATESTDATE &&
+        param2.genericType.value.rawType.name === PRIMITIVE_TYPE.LATESTDATE
+      ) {
+        return true;
+      }
+      return (
+        param1.genericType.value.rawType.name ===
+          param2.genericType.value.rawType.name &&
+        param1.values[0] === param2.values[0]
+      );
+    }
+    return false;
+  };
+  switch (stereotype) {
+    case MILESTONING_STEREOTYPE.BITEMPORAL:
+      return (
+        (idx === 0 &&
+          checkIfEquivalent(parameterValue, milestoningDate.processingDate)) ||
+        (idx === 1 &&
+          checkIfEquivalent(parameterValue, milestoningDate.businessDate))
+      );
+    case MILESTONING_STEREOTYPE.PROCESSING_TEMPORAL:
+      return checkIfEquivalent(parameterValue, milestoningDate.processingDate);
+    case MILESTONING_STEREOTYPE.BUSINESS_TEMPORAL:
+      return checkIfEquivalent(parameterValue, milestoningDate.businessDate);
+    default:
+  }
+  return false;
+};
+
+/**
+ * Checks if INTERNAL_PropagatedValue we passed for milestoned properties can be removed or not.
+ * We try to make analysis on whether date propagations is supported for this particular property or not:
+ * 1. If it is supported we check if the value it has is same as global milestoning dates
+ *    (i) Yes -> We replace it with actual ValueSpecification (We assune that here it is not pointing to the preceeding property so date
+ *        propagation cannot be done).
+ *    (ii) No -> We just remove it.
+ * 2. If not supported we replace it with the actual ValueSpecification.
+ */
+export const checkIfInternalPropagatedValueCanBeRemoved = (
```

Comment:
--> `isPropagatedValueRemovable`

When I think about propagated value (i.e. the bubble), I always think about a generic structure, so I really dislike this kind of method where we have differentiating behavior that pertains to milestoning only like this. What do you think about having a simple boolean flag inside of the propagated value to indicate if we should squash it or pop it. This way, the differentiating logic only lies at where we create the bubble or edit the bubble. When we build value specification, there's no special thing we need to do (I think)